# Exercise 1

In [10]:
import numpy as np
xs = np.array([1.2, 2.2, 2.7, 2.8, 3.1, 3.5])
ys = np.array([0, 0, 1, 0, 1, 1])

def sigmoid(z):
    return 1 / (1 + np.exp(-z))

def predict(xs, w, b):
    return sigmoid(w * xs + b)

def bce_loss(ys, ys_pred, verbose=False):
    losses = -(ys * np.log(ys_pred) + (1 - ys) * np.log(1 - ys_pred))
    if verbose:
        for i in range(len(losses)):
            print(f"Loss for sample {i}: {losses[i]:.4f}")
    return losses.mean()

# (1) Calculate BCE loss for w = 3 and b = -6
w = 3
b = -6
ys_pred = predict(xs, w, b)
loss = bce_loss(ys, ys_pred, verbose=True)
print(f"Loss for w = {w}, b = {b}: {loss:.4f}")

# (2) Calculate BCE loss for w = 4 and b = -10
w = 4
b = -10
ys_pred = predict(xs, w, b)
loss = bce_loss(ys, ys_pred, verbose=True)
print(f"Loss for w = {w}, b = {b}: {loss:.4f}")

Loss for sample 0: 0.0868
Loss for sample 1: 1.0375
Loss for sample 2: 0.1155
Loss for sample 3: 2.4868
Loss for sample 4: 0.0362
Loss for sample 5: 0.0110
Loss for w = 3, b = -6: 0.6290
Loss for sample 0: 0.0055
Loss for sample 1: 0.2633
Loss for sample 2: 0.3711
Loss for sample 3: 1.4633
Loss for sample 4: 0.0868
Loss for sample 5: 0.0181
Loss for w = 4, b = -10: 0.3680


# Exercise 2

In [11]:
def bce_gradient(xs, ys, ys_pred, verbose=False):
    dL_dz = ys_pred - ys
    dL_dw = (dL_dz * xs)
    dL_db = dL_dz
    if verbose:
        for i in range(len(xs)):
            print(f"Gradient for sample {i}: dL/dw = {dL_dw[i]:.4f}, dL/db = {dL_db[i]:.4f}")
    return dL_dw.mean(), dL_db.mean()

# (1) Calculate the gradient for w = 3 and b = -6
w = 3
b = -6
ys_pred = predict(xs, w, b)
dL_dw, dL_db = bce_gradient(xs, ys, ys_pred, verbose=True)
print(f"Gradient for w = {w}, b = {b}: -dL/dw = {-dL_dw:.4f}, -dL/db = {-dL_db:.4f}")

# (2) Calculate the gradient for w = 4 and b = -10
w = 4
b = -10
ys_pred = predict(xs, w, b)
dL_dw, dL_db = bce_gradient(xs, ys, ys_pred, verbose=True)
print(f"Gradient for w = {w}, b = {b}: -dL/dw = {-dL_dw:.4f}, -dL/db = {-dL_db:.4f}")

Gradient for sample 0: dL/dw = 0.0998, dL/db = 0.0832
Gradient for sample 1: dL/dw = 1.4204, dL/db = 0.6457
Gradient for sample 2: dL/dw = -0.2946, dL/db = -0.1091
Gradient for sample 3: dL/dw = 2.5671, dL/db = 0.9168
Gradient for sample 4: dL/dw = -0.1103, dL/db = -0.0356
Gradient for sample 5: dL/dw = -0.0385, dL/db = -0.0110
Gradient for w = 3, b = -6: -dL/dw = -0.6073, -dL/db = -0.2483
Gradient for sample 0: dL/dw = 0.0066, dL/db = 0.0055
Gradient for sample 1: dL/dw = 0.5092, dL/db = 0.2315
Gradient for sample 2: dL/dw = -0.8371, dL/db = -0.3100
Gradient for sample 3: dL/dw = 2.1519, dL/db = 0.7685
Gradient for sample 4: dL/dw = -0.2578, dL/db = -0.0832
Gradient for sample 5: dL/dw = -0.0630, dL/db = -0.0180
Gradient for w = 4, b = -10: -dL/dw = -0.2516, -dL/db = -0.0991


# Exercise 3

In [29]:
x1s = np.array([1, 2, 3, 4, 4, 6])
x2s = np.array([4, 6, 2, 4, 6, 3])
ys = np.array([0, 0, 1, 1, 0, 1])
N = 6

w = [0.0, 0.0]
b = 0.0
lr = 0.1

for epoch in range(10000):
    # (1) Calculate the loss and gradients
    ys_pred = sigmoid(w[0] * x1s + w[1] * x2s + b)
    loss = bce_loss(ys, ys_pred)
    dw0 = (1/N) * np.sum(x1s * (ys_pred - ys))
    dw1 = (1/N) * np.sum(x2s * (ys_pred - ys))
    db = (1/N) * np.sum(ys_pred - ys)
    # (2) Update the parameters
    w[0] -= lr * dw0
    w[1] -= lr * dw1
    b -= lr * db
    # (3) Print the loss and w, b for each epoch
    print(f"epoch: {epoch}, loss: {loss:.4f}, w1: {w[0]:.4f}, w2: {w[1]:.4f}, b: {b:.4f}")
    
print(f"final loss: {loss:.4f}, w1: {w[0]:.4f}, w2: {w[1]:.4f}, b: {b:.4f}")

epoch: 0, loss: 0.6931, w1: 0.0500, w2: -0.0583, b: 0.0000
epoch: 1, loss: 0.6370, w1: 0.1023, w2: -0.1049, b: 0.0019
epoch: 2, loss: 0.5896, w1: 0.1522, w2: -0.1466, b: 0.0042
epoch: 3, loss: 0.5488, w1: 0.1988, w2: -0.1851, b: 0.0067
epoch: 4, loss: 0.5136, w1: 0.2422, w2: -0.2208, b: 0.0091
epoch: 5, loss: 0.4829, w1: 0.2828, w2: -0.2542, b: 0.0116
epoch: 6, loss: 0.4561, w1: 0.3209, w2: -0.2855, b: 0.0141
epoch: 7, loss: 0.4325, w1: 0.3567, w2: -0.3149, b: 0.0166
epoch: 8, loss: 0.4116, w1: 0.3904, w2: -0.3426, b: 0.0191
epoch: 9, loss: 0.3930, w1: 0.4224, w2: -0.3687, b: 0.0216
epoch: 10, loss: 0.3763, w1: 0.4526, w2: -0.3935, b: 0.0241
epoch: 11, loss: 0.3613, w1: 0.4815, w2: -0.4171, b: 0.0266
epoch: 12, loss: 0.3477, w1: 0.5089, w2: -0.4395, b: 0.0291
epoch: 13, loss: 0.3354, w1: 0.5352, w2: -0.4609, b: 0.0316
epoch: 14, loss: 0.3241, w1: 0.5603, w2: -0.4814, b: 0.0341
epoch: 15, loss: 0.3137, w1: 0.5844, w2: -0.5010, b: 0.0365
epoch: 16, loss: 0.3042, w1: 0.6076, w2: -0.5198, 

- lr=3.0: final loss: nan, w1: 10.2543, w2: -8.5386, b: 1.5564
- lr=2.0: final loss: 0.0001, w1: 9.4647, w2: -7.9391, b: 1.7177
- lr=1.0: final loss: 0.0002, w1: 8.5612, w2: -7.2265, b: 1.7689
- lr=0.1: final loss: 0.0024, w1: 5.8354, w2: -4.9553, b: 1.3196
- lr=0.01: final loss: 0.0202, w1: 3.3576, w2: -2.8409, b: 0.6763
- lr=0.001: final loss: 0.1112, w1: 1.4933, w2: -1.2483, b: 0.1918
- lr=0.0001: final loss: 0.3815, w1: 0.4121, w2: -0.3605, b: 0.0216
